keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,474 @@
+- In other cohorts, ClusterQueues might have long running jobs that support
+  checkpointing. In this case, it's necessary to preempt Workloads that are
+  using borrowed quota.
+
+I want to configure in the ClusterQueue how long to wait for Workloads to finish
+up to a limit.
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1alpha2
+kind: ClusterQueue
+metadata:
+  name: cluster-total
+spec:
+  preemption:
+    waitBeforePreemptionSeconds: 600 # leave empty to wait indefinitely.
+```
+
+#### Story 2
+
+As a batch user, some of my workloads implement checkpointing, so I know that
+after preemption they can safely restart. Others need manual intervention before
+restart, so they should be marked as failed.
+
+This can be configured like this:
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1alpha2
+kind: Workload
+metadata:
+  name: my-workload
+spec:
+  onPreemption: Requeue # OR Terminate
+```
+
+### Notes/Contraints/Caveats (Optional)
+
+#### How to configure waitBeforePreemptionSeconds
+
+The proposed field `ClusterQueue.spec.preemption.waitBeforePreemptionSeconds`
+can be interpreted in two ways:
+1. **How long jobs are willing to wait**.
+   This shouldn't be problematic. The field can be configured based purely on
+   the importance of the Workloads served by the preempting ClusterQueue.
+2. **The characteristics of the workloads in the cohort**; for example, how long
+   they take to finish, on average.
+   This implies that all workloads in the cohort have similar characteristics
+   and all the ClusterQueues in the cohort should have the same wait period.
+
+This caveat should be part of the documentation as a best practice for how to
+setup the field.
+
+#### Workload preemption doesn't imply immediate Pod termination
+
+When Kueue issues a Workload preemption, the workload API integration controller
+is expected to start removing Pods.
+In the case of Kubernetes' batch.v1/Job, the following steps happen:
+1. Kueue's job controller sets the
+[`.spec.suspend`](https://kubernetes.io/docs/concepts/workloads/controllers/job/#suspending-a-job)
+field to true.
+2. The Kubernetes' job controller deletes the Job's Pods.
+3. The Kubelets send SIGTERM signals to the Pod's containers, which can have a
+   graceful termination logic.
+
+This implies the following:
+- Pods of a workload could implement checkpointing as part of their graceful
+  termination.
+- The resources from these Pods are not immediately available and they could
+  be arbitrarily delayed.
+- While Pods terminate, a ClusterQueue's quota could be oversubscribed.
+
+Oversubscription could be mitigated by watching the running pods but this will
+be left for future work.
+
+#### Why no control to opt-out a ClusterQueue from preemption
+
+In a cohort, some ClusterQueue could have high priority Workloads running, so
+it might be desired not to disturb them.
+
+However, this could be achieved by two means:
+- Restricting other ClusterQueues to only have workloads with lower priority,
```

Comment:
Maybe add a new field to indicate whether workload can be preempted is a better choice. These two ways are restricting the usages of clusterQueues. But it's actually not one of this KEP goals.

Reply:
Do you have a use case where the knob would not be related to priority?

Reply:
One scenario here:
1. We have two clusterQueues, cqA and cqB. They belong to a same cohort.
3. We create a workloadA in cqA, but cqA is insufficient of resources, so it will lend some resources from cqB
4. Then we create a workloadB in cqB, workloadB's priority is lower than workloadA, but cqB is also lack of resources, so at the moment:
    1）we'll perform preemption although workloadA's priority is higher. It shows that I can share some resources with neighbors, but when I need, I will reclaim.
    2)  we will not preempt workloadA for it's priority is higher, we respect priority whatever.

Will we consider option 1

Reply:
Yes, most reviewers are suggesting to start with a minimal API and improve later based on feedback.